pymc_marketing/mmm/budget_optimizer.py

```diff
@@ -20,6 +21,59 @@
 from pymc_marketing.mmm.transformers import michaelis_menten
 from pymc_marketing.mmm.utils import sigmoid_saturation
 
+class BudgetOptimizer:
+    def __init__(self, adstock, saturation, num_days, channels, adstock_first=True):
+        self.adstock = adstock
+        self.saturation = saturation
+        self.num_days = num_days
+        self.channels = channels
+        self.adstock_first = adstock_first
+
+    def objective(self, budgets):
+        total_response = 0
+        first_transform, second_transform = (
+            (self.adstock, self.saturation) if self.adstock_first else (self.saturation, self.adstock)
+        )
+        for idx, (channel, params) in enumerate(self.channels.items()):
+            budget = budgets[idx]
+            first_params = params['adstock_params'] if self.adstock_first else params['saturation_params']
+            second_params = params['saturation_params'] if self.adstock_first else params['adstock_params']
+            spend = np.full(self.num_days, budget)
+            spend_extended = np.concatenate([spend, np.zeros(self.adstock.l_max)])
+            transformed_spend = second_transform.function(x=first_transform.function(x=spend_extended, **first_params).eval(), **second_params).eval()
```

Comment:
Maybe this should be compiled? Is this slow in the minimize function?

Reply:
I can look into this if you'd like. Or maybe we note and just create an issue for later

Reply:
I feel better to work as MVP, it's not to long based on my test on mmm example!